File: app/src/main/java/com/goel/attendancetracker/activities/BackupRestoreActivity.java
package com.goel.attendancetracker.activities;

import android.app.AlertDialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.goel.attendancetracker.R;
import com.goel.attendancetracker.database.Params;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import java.io.File;
import java.util.Objects;

public class BackupRestoreActivity extends AppCompatActivity {

    public static FirebaseUser user;
    public static final String SIGN_OUT = "com.goel.attendancetracker.signout";
    private StorageReference storageRef;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_backup_restore);
        Objects.requireNonNull(getSupportActionBar()).setTitle("Backup & Restore");
        ((TextView) findViewById(R.id.user_email)).setText(user.getEmail());

        storageRef = FirebaseStorage.getInstance().getReference(user.getUid() + "/" + Params.DB_NAME);
    }

    public void createBackup(View view){
        new AlertDialog.Builder(BackupRestoreActivity.this)
                .setTitle("Confirm Backup")
                .setMessage("Creating the backup will overwrite the existing backup. Do want to continue?")
                .setPositiveButton("Confirm", (dialog, which) -> {
                    dialog.dismiss();
                    backupDatabase();
                })

                .setNegativeButton("Cancel", (dialog, which) -> dialog.dismiss())
                .setIcon(android.R.drawable.ic_dialog_info)
                .show();
    }

    private void backupDatabase(){
        ConstraintLayout loading = findViewById(R.id.backup_restore_loading);
        loading.setVisibility(View.VISIBLE);
        Uri databaseUri = Uri.fromFile(getDatabasePath(Params.DB_NAME));
        UploadTask backupTask = storageRef.putFile(databaseUri);
        backupTask.addOnSuccessListener(taskSnapshot -> {
            Toast toast = Toast.makeText(BackupRestoreActivity.this, "Backed Up Successfully", Toast.LENGTH_LONG);
            toast.setGravity(Gravity.CENTER, 0, 50);
            toast.show();
            loading.setVisibility(View.INVISIBLE);
        });
        backupTask.addOnFailureListener(e -> {
            Toast toast = Toast.makeText(BackupRestoreActivity.this, "Back Up Failed", Toast.LENGTH_LONG);
            toast.setGravity(Gravity.CENTER, 0, 50);
            toast.show();
            loading.setVisibility(View.INVISIBLE);
        });

    }

    public void createRestore(View view){
        new AlertDialog.Builder(BackupRestoreActivity.this)
                .setTitle("Confirm Restore")
                .setMessage("Performing the restore will overwrite the existing data. Do want to continue?")
                .setPositiveButton("Confirm", (dialog, which) -> {
                    dialog.dismiss();
                    restoreDatabase();
                })

                .setNegativeButton("Cancel", (dialog, which) -> dialog.dismiss())
                .setIcon(android.R.drawable.ic_dialog_info)
                .show();
    }

    private void restoreDatabase() {
        ConstraintLayout loading = findViewById(R.id.backup_restore_loading);
        loading.setVisibility(View.VISIBLE);
        File databaseFile = new File(getDatabasePath(Params.DB_NAME).getPath());

        storageRef.getFile(databaseFile)
                .addOnSuccessListener(taskSnapshot -> {
                    Toast toast = Toast.makeText(BackupRestoreActivity.this, "Data Restored Successfully", Toast.LENGTH_LONG);
                    toast.setGravity(Gravity.CENTER, 0, 50);
                    toast.show();
                            loading.setVisibility(View.INVISIBLE);
                })
                .addOnFailureListener(e -> {
                    loading.setVisibility(View.INVISIBLE);
                    if (e.getMessage().equals("Object does not exist at location.")){
                        backupNotFound();
                    }
                    Toast toast = Toast.makeText(BackupRestoreActivity.this, "Restore Failed", Toast.LENGTH_SHORT);
                    toast.setGravity(Gravity.CENTER, 0, 50);
                    toast.show();
                });
    }

    private void backupNotFound() {
        new AlertDialog.Builder(BackupRestoreActivity.this)
                .setTitle("Backup Not Found")
                .setMessage("No backup found on this account. Please try with a different account")
                .setPositiveButton("OK", (dialog, which) -> dialog.dismiss())
                .setIcon(android.R.drawable.ic_dialog_alert)
                .show();
    }

    public void logOut(View view){
        FirebaseAuth.getInstance().signOut();
        startActivity(new Intent(BackupRestoreActivity.this, SignInActivity.class).putExtra(SIGN_OUT, true));
        finish();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        startActivity(new Intent(BackupRestoreActivity.this, MainActivity.class));
        finish();
    }
}

File: app/src/main/java/com/goel/attendancetracker/activities/BackupRestoreActivity.kt
package com.goel.attendancetracker.activities

import android.app.AlertDialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.goel.attendancetracker.R
import com.goel.attendancetracker.databinding.ActivityBackupRestoreBinding
import com.goel.attendancetracker.utils.Constants
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import java.io.File
import java.util.*

class BackupRestoreActivity : AppCompatActivity() {

    private lateinit var binding: ActivityBackupRestoreBinding
    private lateinit var storageRef: StorageReference

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityBackupRestoreBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.title = getString(R.string.backup_and_restore)
        binding.userEmail.text = user.email
        storageRef = FirebaseStorage.getInstance().getReference("${user.uid}/${Constants.DB_NAME}")
        setClickListeners()
    }

    private fun setClickListeners() {
        binding.backupButton.setOnClickListener {
            createBackup()
        }
        binding.restoreBackup.setOnClickListener {
            createRestore()
        }

        binding.logoutButton.setOnClickListener {
            logOut()
        }
    }

    private fun createBackup() {
        AlertDialog.Builder(this@BackupRestoreActivity)
            .setTitle(R.string.confirm_backup)
            .setMessage(R.string.backup_prompt)
            .setPositiveButton(R.string.confirm) { dialog, _ ->
                dialog.dismiss()
                backupDatabase()
            }
            .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
            .setIcon(android.R.drawable.ic_dialog_info)
            .show()
    }

    private fun backupDatabase() {
        val loading = binding.backupRestoreLoading
        loading.visibility = View.VISIBLE
        val databaseUri = Uri.fromFile(getDatabasePath(Constants.DB_NAME))
        val backupTask = storageRef.putFile(databaseUri)
        backupTask.addOnSuccessListener {
            val toast = Toast.makeText(
                this@BackupRestoreActivity,
                R.string.backup_successful,
                Toast.LENGTH_LONG
            )
            toast.setGravity(Gravity.CENTER, 0, 50)
            toast.show()
            loading.visibility = View.INVISIBLE
        }
        backupTask.addOnFailureListener {
            val toast =
                Toast.makeText(this@BackupRestoreActivity, R.string.backup_failed, Toast.LENGTH_LONG)
            toast.setGravity(Gravity.CENTER, 0, 50)
            toast.show()
            loading.visibility = View.INVISIBLE
        }
    }

    private fun createRestore() {
        AlertDialog.Builder(this@BackupRestoreActivity)
            .setTitle(R.string.confirm_restore)
            .setMessage(R.string.restore_prompt)
            .setPositiveButton(R.string.confirm) { dialog, _ ->
                dialog.dismiss()
                restoreDatabase()
            }
            .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
            .setIcon(android.R.drawable.ic_dialog_info)
            .show()
    }

    private fun restoreDatabase() {
        val loading = binding.backupRestoreLoading
        loading.visibility = View.VISIBLE
        val databaseFile = File(getDatabasePath(Constants.DB_NAME).path)
        storageRef.getFile(databaseFile)
            .addOnSuccessListener {
                val toast = Toast.makeText(
                    this@BackupRestoreActivity,
                    R.string.restore_successful,
                    Toast.LENGTH_LONG
                )
                toast.setGravity(Gravity.CENTER, 0, 50)
                toast.show()
                loading.visibility = View.INVISIBLE
            }
            .addOnFailureListener { e: Exception ->
                loading.visibility = View.INVISIBLE
                if (e.message == getString(R.string.backup_not_found_exception)) {
                    backupNotFound()
                }
                val toast =
                    Toast.makeText(this@BackupRestoreActivity, R.string.restore_failed, Toast.LENGTH_SHORT)
                toast.setGravity(Gravity.CENTER, 0, 50)
                toast.show()
            }
    }

    private fun backupNotFound() {
        AlertDialog.Builder(this@BackupRestoreActivity)
            .setTitle(R.string.backup_not_found_title)
            .setMessage(R.string.backup_not_found_message)
            .setPositiveButton(R.string.ok) { dialog, _ -> dialog.dismiss() }
            .setIcon(android.R.drawable.ic_dialog_alert)
            .show()
    }

    private fun logOut() {
        FirebaseAuth.getInstance().signOut()
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestIdToken(getString(R.string.default_web_client_id))
            .requestEmail()
            .build()
        GoogleSignIn.getClient(this, gso).signOut()
            .addOnSuccessListener {
                Toast.makeText(this, R.string.signed_out, Toast.LENGTH_SHORT).show()
                startActivity(Intent(this, SignInActivity::class.java))
                finish()
            }

    }

    override fun onBackPressed() {
        super.onBackPressed()
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }

    companion object {
        @JvmStatic
        lateinit var user: FirebaseUser
    }
}